app/models/user.rb

```diff
@@ -1,2 +1,28 @@
+# frozen_string_literal: true
 class User < ApplicationRecord
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i.freeze
+
+  before_save { email.downcase! }
+
+  validates :first_name,
+            presence: true,
+            length: { maximum: 50 }
+
+  validates :last_name,
+            presence: true,
+            length: { maximum: 50 }
+
+  validates :username,
+            presence: true,
+            uniqueness: { case_sensitive: false },
+            length: { maximum: 15 }
+
+  validates :email,
+            presence: true,
+            uniqueness: true,
+            length: { maximum: 255 },
+            format: { with: VALID_EMAIL_REGEX }
```

Comment:
For tutorial purposes this is OK, but look at using a proper email validation library for real apps.

Reply:
I'll replace this with a library in any case just for the practice, thanks!